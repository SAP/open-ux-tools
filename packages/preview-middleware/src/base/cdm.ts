import { readFileSync } from 'fs';
import { join } from 'path';
import type { TemplateConfig } from './config';
import { type FLPCdmConfig, FLPHomePageDefaults } from '../types';

/**
 * Generates a CDM by embedding the provided app tiles into the FLP homepage.
 *
 * @param apps - A list of app to be embedded.
 * @returns The generated CDM configuration
 */
export function generateCdm(apps: TemplateConfig['apps']): FLPCdmConfig {
    const cdm = JSON.parse(readFileSync(join(__dirname, '../../templates/flp/cdm.base.json'), 'utf-8')) as FLPCdmConfig;

    // add apps
    Object.keys(apps).forEach((id) => {
        const appId = apps[id].additionalInformation.split('=')[1];
        const vizId = `VIZ:${appId}`;
        const [object, action] = id.split('-');
        const { title, description, url } = apps[id];

        // add app to default catalog
        cdm.catalogs[FLPHomePageDefaults.catalogId].payload.viz.push(vizId);

        // create flp visualization
        cdm.visualizations[vizId] = {
            'vizType': 'sap.ushell.StaticAppLauncher',
            'businessApp': appId,
            'vizConfig': {
                'sap.app': {
                    title,
                    subTitle: description
                },
                'sap.flp': {
                    'target': {
                        'appId': appId,
                        'inboundId': `${object}-${action}`,
                        'parameters': [
                            {
                                'name': 'sap-ui-tech-hint',
                                'value': 'UI5'
                            }
                        ]
                    }
                }
            }
        };

        // create flp application
        cdm.applications[appId] = {
            'sap.app': {
                id: appId,
                title,
                crossNavigation: {
                    inbounds: {
                        [`${object}-${action}`]: {
                            'semanticObject': object,
                            'action': action,
                            title,
                            'subTitle': description,
                            'signature': {
                                'additionalParameters': 'allowed'
                            }
                        }
                    }
                }
            },
            'sap.ui5': {
                'componentName': appId
            },
            'sap.ui': {
                'technology': 'UI5'
            },
            'sap.platform.runtime': {
                'componentProperties': {
                    url
                }
            }
        };

        // add app to default section
        cdm.pages[FLPHomePageDefaults.pageName].payload.sections[FLPHomePageDefaults.sectionId].layout.vizOrder.push(
            appId
        );
        cdm.pages[FLPHomePageDefaults.pageName].payload.sections[FLPHomePageDefaults.sectionId].viz[appId] = {
            id: appId,
            vizId
        };
    });

    return cdm;
}
