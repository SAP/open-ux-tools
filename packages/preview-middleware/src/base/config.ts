import { ToolsLogger, type Logger } from '@sap-ux/logger';
import type {
    App,
    DefaultFlpPath,
    DefaultIntent,
    FlpConfig,
    Intent,
    InternalTestConfig,
    MiddlewareConfig,
    TestConfig
} from '../types';
import { render } from 'ejs';
import { join, posix } from 'path';
import { createProjectAccess, getWebappPath, type Manifest, type UI5FlexLayer } from '@sap-ux/project-access';
import { readFileSync } from 'fs';
import { mergeTestConfigDefaults } from './test';
import { type Editor, create } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import type { MergedAppDescriptor } from '@sap-ux/axios-extension';

export interface CustomConnector {
    applyConnector: string;
    writeConnector: string;
    custom: boolean;
}

export interface FlexConnector {
    connector: string;
    layers: string[];
    url?: string;
}

type RtaConfig = NonNullable<MiddlewareConfig['editors']>['rta'];

type TestTemplateConfig = {
    id: string;
    framework: TestConfig['framework'];
    basePath: string;
    initPath: string;
    theme: string;
};

export type PreviewUrls = {
    path: string;
    type: 'preview' | 'editor' | 'test';
};

/**
 * Internal structure used to fill the sandbox.html template
 */
export interface TemplateConfig {
    basePath: string;
    apps: Record<
        string,
        {
            title: string;
            description: string;
            additionalInformation: string;
            applicationType: 'URL';
            url: string;
            applicationDependencies?: MergedAppDescriptor;
        }
    >;
    ui5: {
        libs: string;
        theme: string;
        flex: (CustomConnector | FlexConnector)[];
        bootstrapOptions: string;
        resources: Record<string, string>;
    };
    init?: string;
    flex?: {
        [key: string]: unknown;
        layer: UI5FlexLayer;
        developerMode: boolean;
        pluginScript?: string;
    };
    features?: { feature: string; isEnabled: boolean }[];
    locateReuseLibsScript?: boolean;
}

/**
 * Static settings
 */
export const PREVIEW_URL = {
    client: {
        path: '/preview/client',
        getUrl: (basePath: string) => posix.join(basePath, 'preview', 'client'),
        local: join(__dirname, '../../dist/client'),
        ns: 'open.ux.preview.client'
    },
    api: '/preview/api'
} as const;

/**
 * Default theme
 */
export const DEFAULT_THEME = 'sap_horizon';

/**
 * Default path for mounting the local FLP.
 */
export const DEFAULT_PATH: DefaultFlpPath = '/test/flp.html';

/**
 * Default intent
 */
export const DEFAULT_INTENT = {
    object: 'app',
    action: 'preview'
} as Readonly<DefaultIntent>;

/**
 * SAPUI5 delivered namespaces from https://ui5.sap.com/#/api/sap
 */
const UI5_LIBS = [
    'sap.apf',
    'sap.base',
    'sap.chart',
    'sap.collaboration',
    'sap.f',
    'sap.fe',
    'sap.fileviewer',
    'sap.gantt',
    'sap.landvisz',
    'sap.m',
    'sap.ndc',
    'sap.ovp',
    'sap.rules',
    'sap.suite',
    'sap.tnt',
    'sap.ui',
    'sap.uiext',
    'sap.ushell',
    'sap.uxap',
    'sap.viz',
    'sap.webanalytics',
    'sap.zen'
] as const;

/**
 * Gets the UI5 libs dependencies from manifest.json.
 *
 * @param manifest application manifest
 * @returns UI5 libs that should preloaded
 */
function getUI5Libs(manifest: Partial<Manifest>): string {
    const libs = manifest['sap.ui5']?.dependencies?.libs ?? {};
    // add libs that should always be preloaded
    libs['sap.m'] = {};
    libs['sap.ui.core'] = {};
    libs['sap.ushell'] = {};

    return Object.keys(libs)
        .filter((key) => {
            return UI5_LIBS.some((substring) => {
                return key === substring || key.startsWith(substring + '.');
            });
        })
        .join(',');
}

/**
 * Default configuration for the FLP.
 *
 * @param config partial configuration
 * @returns a full configuration with default values
 */
export function getFlpConfigWithDefaults(config: Partial<FlpConfig> = {}): FlpConfig {
    const flpConfig = {
        path: config.path ?? DEFAULT_PATH,
        intent: config.intent ?? DEFAULT_INTENT,
        apps: config.apps ?? [],
        libs: config.libs,
        theme: config.theme,
        init: config.init
    } satisfies FlpConfig;
    if (!flpConfig.path.startsWith('/')) {
        flpConfig.path = `/${flpConfig.path}`;
    }
    return flpConfig;
}

/**
 * The developer mode is only supported for adaptation projects, therefore, notify the user if it is wrongly configured and then disable it.
 *
 * @param config configurations from the ui5.yaml
 * @param logger logger instance
 */
export function sanitizeConfig(config: MiddlewareConfig, logger: ToolsLogger): void {
    if (config.rta && config.adp === undefined) {
        config.rta.editors = config.rta.editors.map((editor) => {
            if (editor.developerMode) {
                logger.error('developerMode is ONLY supported for SAP UI5 adaptation projects.');
                logger.warn(`developerMode for ${editor.path} disabled`);
                editor.developerMode = false;
            }
            return editor;
        });
    }
}

/**
 * Convert the deprecated RTA configuration.
 *
 * @param rtaDeprecated the deprecated RTA configuration
 * @param rta the RTA configuration
 * @returns the sanitized RTA configuration
 * @private
 */
export function sanitizeRtaConfig(
    rtaDeprecated: MiddlewareConfig['rta'],
    rta: RtaConfig | undefined
): RtaConfig | undefined {
    if (rtaDeprecated) {
        const { editors, ...rta } = rtaDeprecated;
        return { ...rta, endpoints: editors };
    } else {
        return rta;
    }
}

/**
 * Retrieves the configuration settings for UI5 flexibility services.
 *
 * @returns An array of flexibility service configurations, each specifying a connector
 *          and its options, such as the layers it applies to and its service URL, if applicable.
 */
function getFlexSettings(): TemplateConfig['ui5']['flex'] {
    const localConnectorPath = 'custom.connectors.WorkspaceConnector';

    return [
        { connector: 'LrepConnector', layers: [], url: '/sap/bc/lrep' },
        {
            applyConnector: localConnectorPath,
            writeConnector: localConnectorPath,
            custom: true
        },
        {
            connector: 'LocalStorageConnector',
            layers: ['CUSTOMER', 'USER']
        }
    ];
}

/**
 * Add an application to the local FLP preview.
 *
 * @param templateConfig configuration for the preview
 * @param manifest manifest of the additional target app
 * @param app configuration for the preview
 * @param logger logger instance
 */
export async function addApp(
    templateConfig: TemplateConfig,
    manifest: Partial<Manifest>,
    app: App,
    logger: Logger
): Promise<void> {
    const id = manifest['sap.app']?.id ?? '';

    if (manifest['sap.app']?.type === 'component') {
        logger.debug(`No application added for id '${id}' because its type is 'component'`);
        templateConfig.apps = templateConfig.apps ?? {};
        return;
    }

    const appName = getAppName(manifest, app.intent);
    templateConfig.ui5.resources[id] = app.target;
    templateConfig.apps[appName] = {
        title: (await getI18nTextFromProperty(app.local, manifest['sap.app']?.title, logger)) ?? id,
        description: (await getI18nTextFromProperty(app.local, manifest['sap.app']?.description, logger)) ?? '',
        additionalInformation: `SAPUI5.Component=${app.componentId ?? id}`,
        applicationType: 'URL',
        url: app.target
    };
}

/**
 * Get the application name based on the manifest and app configuration.
 *
 * @param manifest - The application manifest.
 * @param intent - The app configuration.
 * @returns The application name.
 */
export function getAppName(manifest: Partial<Manifest>, intent?: Intent): string {
    const id = manifest['sap.app']?.id ?? '';

    intent ??= {
        object: id.replace(/\./g, ''),
        action: 'preview'
    };

    return `${intent?.object}-${intent?.action}`;
}

/**
 * Get the i18n text of the given property.
 *
 * @param projectRoot absolute path to the project root
 * @param propertyValue value of the property
 * @param logger logger instance
 * @returns i18n text of the property
 */
async function getI18nTextFromProperty(
    projectRoot: string | undefined,
    propertyValue: string | undefined,
    logger: Logger
): Promise<string | undefined> {
    //i18n model format could be {{key}} or {i18n>key}
    if (!projectRoot || !propertyValue || propertyValue.search(/{{\w+}}|{i18n>\w+}/g) === -1) {
        return propertyValue;
    }
    const propertyI18nKey = propertyValue.replace(/i18n>|[{}]/g, '');
    const projectAccess = await createProjectAccess(projectRoot);
    const applicationIds = projectAccess.getApplicationIds();
    try {
        const bundle = (await projectAccess.getApplication(applicationIds[0]).getI18nBundles())['sap.app'];
        return bundle[propertyI18nKey]?.[0]?.value?.value ?? propertyI18nKey;
    } catch (e) {
        logger.warn('Failed to load i18n properties bundle');
    }
    return propertyI18nKey;
}

/**
 * Creates the configuration object for the sandbox.html template.
 *
 * @param config FLP configuration
 * @param manifest application manifest
 * @param resources additional resources
 * @returns configuration object for the sandbox.html template
 */
export function createFlpTemplateConfig(
    config: FlpConfig,
    manifest: Partial<Manifest>,
    resources: Record<string, string> = {}
): TemplateConfig {
    const flex = getFlexSettings();
    const supportedThemes: string[] = (manifest['sap.ui5']?.supportedThemes as []) ?? [DEFAULT_THEME];
    const ui5Theme = config.theme ?? (supportedThemes.includes(DEFAULT_THEME) ? DEFAULT_THEME : supportedThemes[0]);
    const id = manifest['sap.app']?.id ?? '';
    const ns = id.replace(/\./g, '/');
    const basePath = posix.relative(posix.dirname(config.path), '/') ?? '.';
    return {
        basePath: basePath,
        apps: {},
        init: config.init ? ns + config.init : undefined,
        ui5: {
            libs: getUI5Libs(manifest),
            theme: ui5Theme,
            flex,
            resources: {
                ...resources,
                [PREVIEW_URL.client.ns]: PREVIEW_URL.client.getUrl(basePath)
            },
            bootstrapOptions: ''
        },
        locateReuseLibsScript: config.libs
    } satisfies TemplateConfig;
}

/**
 * Creates the configuration object for the test template.
 *
 * @param config test configuration
 * @param id application id
 * @param theme theme to be used
 * @returns configuration object for the test template
 */
export function createTestTemplateConfig(config: InternalTestConfig, id: string, theme: string): TestTemplateConfig {
    return {
        id,
        framework: config.framework,
        basePath: posix.relative(posix.dirname(config.path), '/') ?? '.',
        initPath: posix.relative(posix.dirname(config.path), config.init),
        theme
    } satisfies TestTemplateConfig;
}

/**
 * Returns the preview paths.
 *
 * @param config configuration from the ui5.yaml
 * @param logger logger instance
 * @returns an array of preview paths
 */
export function getPreviewPaths(config: MiddlewareConfig, logger: ToolsLogger = new ToolsLogger()): PreviewUrls[] {
    const urls: PreviewUrls[] = [];
    // remove incorrect configurations
    sanitizeConfig(config, logger);
    // add flp preview url
    const flpConfig = getFlpConfigWithDefaults(config.flp);
    urls.push({ path: `${flpConfig.path}#${flpConfig.intent.object}-${flpConfig.intent.action}`, type: 'preview' });
    // add editor urls
    if (config.rta?.editors) {
        config.rta.editors.forEach((editor) => {
            urls.push({ path: editor.path, type: 'editor' });
        });
    }
    // add test urls if configured
    if (config.test) {
        config.test.forEach((test) => {
            const testConfig = mergeTestConfigDefaults(test);
            urls.push({ path: testConfig.path, type: 'test' });
        });
    }
    return urls;
}

const TEMPLATE_PATH = join(__dirname, '../../templates');

/**
 * Generate test runners.
 *
 * @param configs array of test configurations
 * @param manifest application manifest
 * @param fs mem fs editor instance
 * @param webappPath webapp path
 * @param flpTemplConfig FLP configuration
 */
function generateTestRunners(
    configs: TestConfig[] | undefined,
    manifest: Manifest,
    fs: Editor,
    webappPath: string,
    flpTemplConfig: TemplateConfig
): void {
    for (const test of configs ?? []) {
        const testConfig = mergeTestConfigDefaults(test);
        if (['QUnit', 'OPA5'].includes(test.framework)) {
            const testTemlpate = readFileSync(join(TEMPLATE_PATH, 'test/qunit.html'), 'utf-8');
            const testTemplateConfig = createTestTemplateConfig(
                testConfig,
                manifest['sap.app'].id,
                flpTemplConfig.ui5.theme
            );
            fs.write(join(webappPath, testConfig.path), render(testTemlpate, testTemplateConfig));
        } else if (test.framework === 'Testsuite') {
            const testTemlpate = readFileSync(join(TEMPLATE_PATH, 'test/testsuite.qunit.html'), 'utf-8');
            const testTemplateConfig = {
                basePath: flpTemplConfig.basePath,
                initPath: testConfig.init
            };
            fs.write(join(webappPath, testConfig.path), render(testTemlpate, testTemplateConfig));
        }
    }
}

/**
 * Generates the preview files.
 *
 * @param basePath path to the application root
 * @param config configuration from the ui5.yaml
 * @param fs file system editor
 * @param logger logger instance
 * @returns a mem-fs editor with the preview files
 */
export async function generatePreviewFiles(
    basePath: string,
    config: MiddlewareConfig,
    fs?: Editor,
    logger: ToolsLogger = new ToolsLogger()
): Promise<Editor> {
    // remove incorrect configurations
    sanitizeConfig(config, logger);

    // create file system if not provided
    if (!fs) {
        fs = create(createStorage());
    }

    // generate FLP configuration
    const flpTemplate = readFileSync(join(TEMPLATE_PATH, 'flp/sandbox.html'), 'utf-8');
    const flpConfig = getFlpConfigWithDefaults(config.flp);

    const webappPath = await getWebappPath(basePath, fs);
    let manifest: Manifest | undefined;
    if (fs.exists(join(webappPath, 'manifest.json'))) {
        manifest = (await fs.readJSON(join(webappPath, 'manifest.json'))) as unknown as Manifest;
    }
    let flpTemplConfig: TemplateConfig;
    let flpPath: string;

    if (manifest) {
        flpTemplConfig = createFlpTemplateConfig(flpConfig, manifest);
        flpPath = join(webappPath, flpConfig.path);
        await addApp(
            flpTemplConfig,
            manifest,
            {
                target: flpTemplConfig.basePath,
                local: '.',
                intent: flpConfig.intent
            },
            logger
        );
        generateTestRunners(config.test, manifest, fs, webappPath, flpTemplConfig);
    } else {
        flpTemplConfig = createFlpTemplateConfig(flpConfig, {});
        flpPath = join(basePath, flpConfig.path);
    }

    if (flpConfig.apps.length > 0) {
        for (const app of flpConfig.apps) {
            if (app.local) {
                const appPath = await getWebappPath(join(basePath, app.local), fs);
                if (fs.exists(join(appPath, 'manifest.json'))) {
                    const appManifest = (await fs.readJSON(join(appPath, 'manifest.json'))) as unknown as Manifest;
                    await addApp(flpTemplConfig, appManifest, app, logger);
                } else {
                    logger.warn(`Could not add route for ${app}`);
                }
            }
        }
    }
    fs.write(flpPath, render(flpTemplate, flpTemplConfig));

    return fs;
}
