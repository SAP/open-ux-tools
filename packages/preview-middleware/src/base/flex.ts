import type { Logger } from '@sap-ux/logger';
import type { ReaderCollection } from '@ui5/fs';
import { existsSync, mkdirSync, readdirSync, unlinkSync, writeFileSync } from 'fs';
import { join, parse } from 'path';

/**
 * Structure of a flex change.
 */
export interface FlexChange {
    [key: string]: string | object | undefined;
    changeType: string;
    reference: string;
    moduleName?: string;
    content?: {
        codeRef?: string;
        fragmentPath?: string;
    };
}

/**
 * Map of change type specific correction functions.
 */
const moduleNameContentMap: { [key: string]: (change: FlexChange) => string } = {
    codeExt: (change) => (change.content?.codeRef ?? '').replace('.js', ''),
    addXML: (change) => change.content?.fragmentPath ?? ''
};

/**
 * Sets the moduleName property of the provided change to also support old changes with newer UI5 versions.
 *
 * @param change change to be fixed
 * @param logger logger instance
 */
function tryFixChange(change: FlexChange, logger: Logger) {
    try {
        const prefix = change.reference.replace(/\./g, '/');
        change.moduleName = `${prefix}/changes/${moduleNameContentMap[change.changeType](change)}`;
    } catch (error) {
        logger.warn('Could not fix missing module name.');
    }
}

/**
 * Read changes from the file system and return them.
 *
 * @param project reference to the UI5 project
 * @param logger logger instance
 * @returns object with the file name as key and the file content as value
 */
export async function readChanges(project: ReaderCollection, logger: Logger): Promise<Record<string, FlexChange>> {
    const changes: Record<string, FlexChange> = {};
    const files = await project.byGlob('/**/changes/*.*');
    for (const file of files) {
        try {
            const change = JSON.parse(await file.getString()) as FlexChange;
            if (moduleNameContentMap[change.changeType] && !change.moduleName) {
                tryFixChange(change, logger);
            }
            changes[`sap.ui.fl.${parse(file.getName()).name}`] = change;
            logger.debug(`Read change from ${file.getPath()}`);
        } catch (error) {
            logger.warn(error.message);
        }
    }
    return changes;
}

/**
 * Writes flex changes to the user's workspace.
 *
 * @param data JS object to be stored as change
 * @param data.fileName file name that is required for a valid change
 * @param data.fileType file type that is required
 * @param webappPath path to the webapp folder
 * @param logger logger instance
 * @returns object with success flag and optional message
 */
export function writeChange(
    data: object & { fileName?: string; fileType?: string },
    webappPath: string,
    logger: Logger
): { success: boolean; message?: string } {
    const fileName = data.fileName;
    const fileType = data.fileType;
    if (fileName && fileType) {
        logger.debug(`Write change ${fileName}.${fileType}`);
        const path = join(webappPath, 'changes');
        if (!existsSync(path)) {
            mkdirSync(path);
        }
        const filePath = join(path, fileName + '.' + fileType);
        writeFileSync(filePath, JSON.stringify(data, null, 2));
        const message = `FILE_CREATED ${fileName}.${fileType}`;
        return { success: true, message };
    } else {
        return { success: false };
    }
}

/**
 * Deletes an existing flex change from the file system if it exists.
 *
 * @param data JS object of the change to be deleted
 * @param data.fileName file name that is required for a valid change
 * @param webappPath path to the webapp folder
 * @param logger logger instance
 * @returns object with success flag and optional message
 */
export function deleteChange(
    data: object & { fileName?: string },
    webappPath: string,
    logger: Logger
): { success: boolean; message?: string } {
    const fileName = data.fileName?.replace('sap.ui.fl.', '');
    if (fileName) {
        const path = join(webappPath, 'changes');
        if (existsSync(path)) {
            const files = readdirSync(path);
            const file = files.find((element) => element.includes(fileName));
            if (file) {
                logger.debug(`Write change ${file}`);
                const filePath = join(path, file);
                unlinkSync(filePath);
                return { success: true, message: `FILE_DELETED ${file}` };
            }
        }
    }
    return { success: false };
}
