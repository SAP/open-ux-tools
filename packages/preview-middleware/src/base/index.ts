export { FlpSandbox, initAdp } from './flp';
