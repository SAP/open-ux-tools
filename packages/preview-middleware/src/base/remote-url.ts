import { isAppStudio } from '@sap-ux/btp-utils';
import { networkInterfaces } from 'os';
import { devspace } from '@sap/bas-sdk';
import type { ToolsLogger } from '@sap-ux/logger';
import QRCode from 'qrcode';

/**
 * Log remote URL for mobile device access.
 *
 * @param logger ToolsLogger instance
 */
export async function logRemoteUrl(logger: ToolsLogger): Promise<void> {
    try {
        const remoteUrl = await getRemoteUrl(logger);

        const generateQRCode = async (text: string): Promise<void> => {
            try {
                const qrString = await QRCode.toString(text, { type: 'terminal', small: true });
                logger.info(qrString);
            } catch (err) {
                logger.error(err);
            }
        };

        if (remoteUrl) {
            logger.info(`Remote URL: ${remoteUrl}`);
            logger.info('Scan the QR code below with your mobile device to access the preview:');
            await generateQRCode(remoteUrl);
        }
    } catch (error) {
        logger.debug(`Could not generate remote URL: ${error.message}`);
    }
}

/**
 * Get the remote URL for mobile device access.
 *
 * @param logger ToolsLogger instance
 * @returns The remote URL or undefined if not available
 */
export async function getRemoteUrl(logger: ToolsLogger): Promise<string | undefined> {
    try {
        if (isAppStudio()) {
            return await getBASRemoteUrl(logger);
        } else {
            return getVSCodeRemoteUrl(logger);
        }
    } catch (error) {
        logger.error(`Failed to generate remote URL: ${error.message}`);
        return undefined;
    }
}

/**
 * Get remote URL for BAS environment using BAS SDK.
 *
 * @param logger  ToolsLogger instance instance
 * @returns The remote URL from BAS SDK
 */
async function getBASRemoteUrl(logger: ToolsLogger): Promise<string | undefined> {
    try {
        const devspaceInfo = await devspace.getDevspaceInfo();
        if (devspaceInfo?.url) {
            const port = getPortFromArgs() ?? 8080;
            const baseUrl = `${devspaceInfo.url}:${port}`;
            const remoteUrl = appendOpenPath(baseUrl, getOpenPathFromArgs());
            logger.debug(`BAS remote URL generated: ${remoteUrl}`);
            return remoteUrl;
        }
        logger.warn('Could not retrieve BAS remote URL from devspace info');
        return undefined;
    } catch (error) {
        logger.error(`Failed to get BAS remote URL: ${error.message}`);
        return undefined;
    }
}

/**
 * Get remote URL for VSCode environment by detecting network IP.
 * Only works if --accept-remote-connections is enabled.
 *
 * @param logger  ToolsLogger instance instance
 * @returns The remote URL based on network IP
 */
function getVSCodeRemoteUrl(logger: ToolsLogger): string | undefined {
    try {
        const networkIP = getNetworkIP();
        if (!networkIP) {
            logger.warn('Could not determine network IP address for remote access');
            return undefined;
        }
        const protocol = 'http';
        const port = getPortFromArgs() ?? 8080;
        const baseUrl = `${protocol}://${networkIP}:${port}`;
        const remoteUrl = appendOpenPath(baseUrl, getOpenPathFromArgs());

        logger.debug(`VSCode remote URL generated: ${remoteUrl}`);
        return remoteUrl;
    } catch (error) {
        logger.error(`Failed to generate VSCode remote URL: ${error.message}`);
        return undefined;
    }
}

/**
 * Get the first available network IP address (excluding localhost).
 *
 * @returns The network IP address
 */
function getNetworkIP(): string | undefined {
    const interfaces = networkInterfaces();
    // Priority order: prefer non-internal IPv4 addresses
    for (const interfaceName of Object.keys(interfaces)) {
        const networkInterface = interfaces[interfaceName];
        if (!networkInterface) {
            continue;
        }
        for (const alias of networkInterface) {
            // Skip internal (localhost) addresses and IPv6
            if (alias.family === 'IPv4' && !alias.internal) {
                return alias.address;
            }
        }
    }
    return undefined;
}

/**
 * Append the open path to a base URL if provided.
 *
 * @param baseUrl The base URL
 * @param openPath The path to append
 * @returns The complete URL with path appended
 */
function appendOpenPath(baseUrl: string, openPath?: string): string {
    if (!openPath) {
        return baseUrl;
    }
    // Ensure the path starts with a forward slash
    const normalizedPath = openPath.startsWith('/') ? openPath : `/${openPath}`;
    return `${baseUrl}${normalizedPath}`;
}

/**
 * Check if remote connections should be enabled based on command line arguments.
 *
 * @returns Whether --accept-remote-connections is present in process arguments
 */
export function isRemoteConnectionsEnabled(): boolean {
    return process.argv.includes('--accept-remote-connections');
}

/**
 * Extract port from command line arguments or environment.
 *
 * @returns The port number if found
 */
export function getPortFromArgs(): number | undefined {
    // Check for --port argument
    const portIndex = process.argv.findIndex((arg) => arg === '--port' || arg.startsWith('--port='));
    if (portIndex !== -1) {
        const portArg = process.argv[portIndex];
        if (portArg.includes('=')) {
            // --port=8080 format
            const port = parseInt(portArg.split('=')[1], 10);
            return isNaN(port) ? undefined : port;
        } else if (portIndex + 1 < process.argv.length) {
            // --port 8080 format
            const port = parseInt(process.argv[portIndex + 1], 10);
            return isNaN(port) ? undefined : port;
        }
    }

    // Check environment variable
    const envPort = process.env.PORT;
    if (envPort) {
        const port = parseInt(envPort, 10);
        return isNaN(port) ? undefined : port;
    }
    return undefined;
}

/**
 * Extract open path from command line arguments.
 *
 * @returns The path from --open or -o parameter if found
 */
export function getOpenPathFromArgs(): string | undefined {
    // Check for --open or -o argument
    const openIndex = process.argv.findIndex(
        (arg) => arg === '--open' || arg === '-o' || arg.startsWith('--open=') || arg.startsWith('--o=')
    );
    if (openIndex !== -1) {
        const openArg = process.argv[openIndex];
        if (openArg.includes('=')) {
            // --open=path or -o=path format
            return openArg.split('=')[1];
        } else if (openIndex + 1 < process.argv.length) {
            // --open path or -o path format
            return process.argv[openIndex + 1];
        }
    }
    return undefined;
}
