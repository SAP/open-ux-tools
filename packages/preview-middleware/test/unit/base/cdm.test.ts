import { generateCdm } from '../../../src/base/cdm';
import type { TemplateConfig } from '../../../src/base/config';

describe('cdm', () => {
    test('generate cdm without any apps', () => {
        const cdm = generateCdm();
        expect(cdm).toMatchSnapshot();
    });

    test('generate cdm with additional apps', () => {
        const apps = {
            'my-app': {
                title: 'My App',
                url: '/my/app',
                applicationType: 'URL',
                description: 'My App Description',
                additionalInformation: 'SAPUI5.Component=my.app'
            },
            'my-app2': {
                title: 'My App 2',
                url: '/my/app2',
                applicationType: 'URL',
                description: 'My App Description 2',
                additionalInformation: 'SAPUI5.Component=my.app2'
            }
        } as TemplateConfig['apps'];

        const cdm = generateCdm(apps);
        expect(cdm).toMatchSnapshot();
    });
});
