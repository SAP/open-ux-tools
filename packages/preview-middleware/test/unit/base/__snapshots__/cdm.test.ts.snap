// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdm generate cdm with additional apps 1`] = `
Object {
  "_version": "3.1.0",
  "applications": Object {
    "my.app": Object {
      "sap.app": Object {
        "crossNavigation": Object {
          "inbounds": Object {
            "my-app": Object {
              "action": "app",
              "semanticObject": "my",
              "signature": Object {
                "additionalParameters": "allowed",
              },
              "subTitle": "My App Description",
              "title": "My App",
            },
          },
        },
        "id": "my.app",
        "title": "My App",
      },
      "sap.platform.runtime": Object {
        "componentProperties": Object {
          "url": "/my/app",
        },
      },
      "sap.ui": Object {
        "technology": "UI5",
      },
      "sap.ui5": Object {
        "componentName": "my.app",
      },
    },
    "my.app2": Object {
      "sap.app": Object {
        "crossNavigation": Object {
          "inbounds": Object {
            "my-app2": Object {
              "action": "app2",
              "semanticObject": "my",
              "signature": Object {
                "additionalParameters": "allowed",
              },
              "subTitle": "My App Description 2",
              "title": "My App 2",
            },
          },
        },
        "id": "my.app2",
        "title": "My App 2",
      },
      "sap.platform.runtime": Object {
        "componentProperties": Object {
          "url": "/my/app2",
        },
      },
      "sap.ui": Object {
        "technology": "UI5",
      },
      "sap.ui5": Object {
        "componentName": "my.app2",
      },
    },
  },
  "catalogs": Object {
    "homeCatalog": Object {
      "identification": Object {
        "id": "homeCatalog",
        "title": "Homepage Apps",
      },
      "payload": Object {
        "viz": Array [
          "VIZ:my.app",
          "VIZ:my.app2",
        ],
      },
    },
  },
  "pages": Object {
    "SAP_BASIS_PG_UI_MYHOME": Object {
      "identification": Object {
        "id": "SAP_BASIS_PG_UI_MYHOME",
        "title": "My Home",
      },
      "payload": Object {
        "layout": Object {
          "sectionOrder": Array [
            "homeAppsSection",
          ],
        },
        "sections": Object {
          "homeAppsSection": Object {
            "default": true,
            "id": "homeAppsSection",
            "layout": Object {
              "vizOrder": Array [
                "my.app",
                "my.app2",
              ],
            },
            "title": "Recently Added Apps",
            "viz": Object {
              "my.app": Object {
                "id": "my.app",
                "vizId": "VIZ:my.app",
              },
              "my.app2": Object {
                "id": "my.app2",
                "vizId": "VIZ:my.app2",
              },
            },
          },
        },
      },
    },
  },
  "visualizations": Object {
    "VIZ:my.app": Object {
      "businessApp": "my.app",
      "vizConfig": Object {
        "sap.app": Object {
          "subTitle": "My App Description",
          "title": "My App",
        },
        "sap.flp": Object {
          "target": Object {
            "appId": "my.app",
            "inboundId": "my-app",
            "parameters": Array [
              Object {
                "name": "sap-ui-tech-hint",
                "value": "UI5",
              },
            ],
          },
        },
      },
      "vizType": "sap.ushell.StaticAppLauncher",
    },
    "VIZ:my.app2": Object {
      "businessApp": "my.app2",
      "vizConfig": Object {
        "sap.app": Object {
          "subTitle": "My App Description 2",
          "title": "My App 2",
        },
        "sap.flp": Object {
          "target": Object {
            "appId": "my.app2",
            "inboundId": "my-app2",
            "parameters": Array [
              Object {
                "name": "sap-ui-tech-hint",
                "value": "UI5",
              },
            ],
          },
        },
      },
      "vizType": "sap.ushell.StaticAppLauncher",
    },
  },
}
`;

exports[`cdm generate cdm without any apps 1`] = `
Object {
  "_version": "3.1.0",
  "applications": Object {},
  "catalogs": Object {
    "homeCatalog": Object {
      "identification": Object {
        "id": "homeCatalog",
        "title": "Homepage Apps",
      },
      "payload": Object {
        "viz": Array [],
      },
    },
  },
  "pages": Object {
    "SAP_BASIS_PG_UI_MYHOME": Object {
      "identification": Object {
        "id": "SAP_BASIS_PG_UI_MYHOME",
        "title": "My Home",
      },
      "payload": Object {
        "layout": Object {
          "sectionOrder": Array [
            "homeAppsSection",
          ],
        },
        "sections": Object {
          "homeAppsSection": Object {
            "default": true,
            "id": "homeAppsSection",
            "layout": Object {
              "vizOrder": Array [],
            },
            "title": "Recently Added Apps",
            "viz": Object {},
          },
        },
      },
    },
  },
  "visualizations": Object {},
}
`;
